quadratic-core/src/grid/sheets.rs

```
@@ -0,0 +1,307 @@
+        let mut previous: Option<&Sheet> = None;
+        for sheet in self.sheets.iter() {
+            if sheet.id == sheet_id {
+                return previous.map(|previous| previous.order.clone());
+            }
+            previous = Some(sheet);
+        }
+        None
+    }
+
+    pub fn next_sheet(&self, sheet_id: SheetId) -> Option<&Sheet> {
+        let mut next = false;
+        for sheet in self.sheets.iter() {
+            if next {
+                return Some(sheet);
+            }
+            if sheet.id == sheet_id {
+                next = true;
+            };
+        }
+        None
+    }
+
+    /// Adds a sheet to the grid. Returns an error if the sheet name is already
+    /// in use.
+    ///
+    /// If `sheet` is `None`, a new sheet is created with a random ID and
+    /// autogenerated name.
+    ///
+    /// Errors if the sheet name is already in use.
+    pub fn add_sheet(&mut self, sheet: Option<Sheet>) -> Result<SheetId> {
+        // for new sheets, order is after the last one
+        let sheet = sheet.unwrap_or_else(|| {
+            Sheet::new(
+                SheetId::new(),
+                format!("Sheet {}", self.sheets.len() + 1),
+                self.end_order(),
+            )
+        });
+
+        // TODO: handle duplicate ID (probably error)
+
+        // error if duplicate name
+        let id = sheet.id;
+        if self
+            .sheets
+            .iter()
+            .any(|old_sheet| old_sheet.name == sheet.name)
+        {
+            bail!("sheet name already in use");
+        }
+        self.sheets.push(sheet);
+        self.sort_sheets();
+        Ok(id)
+    }
+
+    pub fn remove_sheet(&mut self, sheet_id: SheetId) -> Sheet {
+        let i = self.sheet_id_to_index(sheet_id);
+        match i {
+            Some(i) => self.sheets.remove(i),
+            None => unreachable!(
+                "remove_sheet should never be called with a sheet_id that doesn't exist"
```

It might be better to return an error than panic here.

---

This is actually wrong now.